### PR TITLE
Major refactor to a fully typed api

### PR DESCRIPTION
**Pull request by avive**
Hi guys, here's a first pass on the API. All methods are tested and all tests should pass.
Let's figure out if this is a right direction for the api design. 
The API is now written in typescript and the output is a javascript lib + typescripts typings.
There are still few places that use any but I'm not sure about an easy way around that.
For example, as the wasm exports have no type info, the only way I found to call them from typescript is via `(codec!.exports as any)[funcName](buf)`. 
To look at the types, `yarn build` and have a look at `index.d.ts` in `dist`.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -12,45 +10,169 @@ export async function init(code: BufferSource) {
...
+    return result;
+}
+
+
+///// Internal help functions below
+
 // Call an svm_codec function with the provided buffer. Returns result buffer.
-export function call(funcName: string, buf) {
+function call(funcName: string, buf) : Uint8Array {
+    console.log("caaling svm function: " + funcName);
```
Comment (YaronWittenstein) [inline src/index.ts]: typo

Comment (YaronWittenstein) [reply src/index.ts]: please change to SVM (capital letters)

File: src/index.ts [hunk 2]
```diff
@@ -59,74 +181,62 @@ export function wasmBufferDataSlice(buf, offset, length: number) {
...
 
     return JSON.parse(string);
 }
 
 // Returns a utf-8 string representation of an error in an svm_codec buffer
-export function loadWasmBufferError(buf) {
+function loadWasmBufferError(buf) {
     let length = wasmBufferLength(buf);
     const slice = wasmBufferDataSlice(buf, 0, length);
     // assert.strictEqual(slice[0], ERR_MARKER);
```
Comment (YaronWittenstein) [inline src/index.ts]: I think we want that `assert`

File: src/index.ts [hunk 3]
```diff
@@ -12,45 +10,169 @@ export async function init(code: BufferSource) {
     codec = await WebAssembly.instantiate(wasm, {})
 }
 
+// Returns true iff library was initialized with wasm code
```
Comment (YaronWittenstein) [inline src/index.ts]: Wasm

File: src/index.ts [hunk 4]
```diff
@@ -12,45 +10,169 @@ export async function init(code: BufferSource) {
...
+    version: number,
+    template: string,
+    name: string
+    ctor_name: string,
+    calldata: Uint8Array,
+}
+
+// User input to the encodeCallData function
+export interface CallData {
+    version: number,
```
Comment (YaronWittenstein) [inline src/index.ts:29]: basically, the `version` should always be 0 for now
do we want to have that part of the `interface`?

**Review by YaronWittenstein — CHANGES_REQUESTED** [review]
Gave a few minor comments.
Overall, looks good!

File: src/index.ts (hunk 1 above)
Comment (YaronWittenstein) [reply src/index.ts]: @avive

File: src/index.ts (hunk 4 above)
Comment (YaronWittenstein) [reply src/index.ts:29]: @avive

File: src/index.ts (hunk 2 above)
Comment (YaronWittenstein) [reply src/index.ts]: @avive

File: tests/index.test.ts [hunk 5]
```diff
@@ -327,15 +194,8 @@ describe("Call Account", function () {
...
-        const buf = SvmCodec.newWasmBuffer(tx);
-        const result = SvmCodec.call("wasm_encode_call", buf);
-        const error = SvmCodec.loadWasmBufferError(result);
-        expect(error).toStrictEqual(
-            "The value of a specific field is invalid (`target`)."
-        );
+        expect(() => encodeCall("102030", new Uint8Array(), new Uint8Array()))
+            .toThrow("The value of a specific field is invalid (`target`).");
 
-        SvmCodec.wasmBufferFree(buf);
```
Comment (YaronWittenstein) [inline tests/index.test.ts:338]: It's a good practice to demonstrate proper usage of the API.
If the `svm_codec.wasm` is a long-lived Wasm instance,
it's important not to leak memory.

**Review by YaronWittenstein — CHANGES_REQUESTED** [review]
Please review all my previous comments prior to merging this PR

File: src/index.ts (hunk 3 above)
Comment (YaronWittenstein) [reply src/index.ts]: @avive

File: src/index.ts (hunk 1 above)
Comment (avive) [reply src/index.ts]: I removed this debug line in the latest commit.

File: src/index.ts (hunk 2 above)
Comment (avive) [reply src/index.ts]: Ok but it should return in error string in this case and not throw. Something like this:
if (slice[0] !== ERR_MARKER) {
        return "svm buffer doesn't have an error."
}
return new TextDecoder().decode(slice.slice(1));

File: src/index.ts (hunk 3 above)
Comment (avive) [reply src/index.ts]: done

File: tests/index.test.ts (hunk 5 above)
Comment (avive) [reply tests/index.test.ts:338]: There are no frees in the test (mock client code) - only in the api implementation itself where memory is allocated. So there is no api to free buffers right now - are you saying we should expose the free buffer method?

File: src/index.ts [hunk 6]
```diff
@@ -1,132 +1,231 @@
...
+    target: string,
+    func_name: string,
+    verifydata: Uint8Array,
+    calldata: Uint8Array,
+}
+
+
+// Encodes the provided spawn app data
+export function encodeSpawn(data: Spawn) : Uint8Array {
```
Comment (YaronWittenstein) [inline src/index.ts:38]: Both the `encodeSpawn`, `encodeCall` and `encodeInput` behave the same
(`wasm_encode_spawn` vs `wasm_encode_call` vs `wasm_encode_inputdata`)

I think you can extract both into a common abstraction. 
Maybe calling it `encodeMessage` would make a good naming.

Later in the review, I've given a similar comment for the decoding side.

File: src/index.ts [hunk 7]
```diff
@@ -1,132 +1,231 @@
...
+
+    if (errMessage.length > 0) {
+        throw errMessage;
+    }
+
+    return slice.slice(1);
+}
+
+// Decodes the SVM encoded call account data
+export function decodeCall(bytes: Uint8Array): JSON {
```
Comment (YaronWittenstein) [inline src/index.ts:86]: Both `decodeCall`, `decodeSpawn` and `decodeInput` behave pretty much the same. 
The only difference is the function name passed the `call.`

Probably we'd be better with extracting the shared code into a `decodeMessage` function.

**Review by YaronWittenstein — CHANGES_REQUESTED** [review]

File: tests/index.test.ts (hunk 5 above)
Comment (avive) [reply tests/index.test.ts:338]: I've made the free function public and the test code need to change to free buffers returned from the api to the client for educational purposes.

File: src/index.ts (hunk 7 above)
Comment (avive) [reply src/index.ts:86]: I don't think that further refactoring should be part of this PR. It should be done IMO in a follow-up PR. There's always more to optimize and improve but I think that the basic impl now is good enough for v0.1 and so smapp 0.3 can integrate it. We'll learn much more once the first api client is actually trying to use it in real-world use-cases and obviously much more is going to need to change. Maybe it is best to optimize once such changes were made.

**Review by YaronWittenstein — APPROVED** [review]

File: src/index.ts [hunk 8]
```diff
@@ -1,132 +1,231 @@
...
-
 export const OK_MARKER = 1;
 export const ERR_MARKER = 0;
 
 let codec: WebAssembly.Instance;
 
 // Init the library with the binary content of a .wasm file.
 // This function must be called once per client prior to use of any other package functions.
-export async function init(code: BufferSource) {
+export async function init(code: BufferSource) : Promise<void> {
```
Comment (brusherru) [inline src/index.ts:8]: Why don't return `Promise<WebAssembly.Instance>` instead of mutating some variable?
Or, moreover, return a "ready-to-use" module with functions, so there will be no chance of trying to call something on an uninstantiated wasm instance.

E.G.
```
export async function init(code: BufferSource) => {
  const wasm = await WebAssembly.compile(code);
  const codec = await WebAssembly.instantiate(wasm, {});
  const call = (funcName: string, buf) : Uint8Array => (codec.exports as any)[funcName](buf);
  return {
    encodeSpawn: ...,
    ...
  };
}
```

In case that you want to provide easy access to API from any place in the code without a requirement for the developer to store the instance somewhere (aka in global scope) and pass through multiple files, and there can be only one instance of wasm (as it done here), you can use Singleton pattern. The main cons, it will infect all functions with Promises.

File: src/index.ts [hunk 9]
```diff
@@ -1,132 +1,231 @@
...
+            s = "0" + s;
+        }
+        result += s;
+    }
+    return result;
+}
+
+// Call an svm_codec function with the provided buffer. Returns result buffer.
+function call(funcName: string, buf) : Uint8Array {
+    return (codec.exports as any)[funcName](buf)
```
Comment (brusherru) [inline src/index.ts:142]: To follow up my first comment, such functions can fail with `Cannot read property ... of undefined` in case that someone called it without instantiating wasm instance first.
You have to add a check for `isInstantiate` here and throw a reasonable error or choose another approach to instantiate and provide functions to the developer.

File: src/index.ts [hunk 10]
```diff
@@ -1,132 +1,231 @@
...
+
+
+// Encodes the provided spawn app data
+export function encodeSpawn(data: Spawn) : Uint8Array {
+    const buf = newWasmBuffer(data);
+    const result = call("wasm_encode_spawn", buf);
+    const len = wasmBufferLength(result);
+    const slice = wasmBufferDataSlice(result, 0, len);
+
+    let errMessage = "";
```
Comment (brusherru) [inline src/index.ts]: I'd prefer to make it more straight-forward:
```
const errMessage = slice[0] !== OK_MARKER ? loadWasmBufferError(result) : null;
```
(or empty string instead of null to keep it always be `string` type)

File: src/index.ts [hunk 11]
```diff
@@ -1,132 +1,231 @@
...
-        console.warn(msg);
-        throw msg;
+        throw loadWasmBufferError(buf);
     }
 
-    assert.strictEqual(slice[0], OK_MARKER);
-
     const string = loadWasmBufferDataAsString(buf);
 
     return JSON.parse(string);
```
Comment (brusherru) [inline src/index.ts:229]: It can fail with something like `Unexpected token X on position N`.
I'd prefer to make a more descriptive error message by wrapping it with `try-catch` and provide also input string to the error message.

File: src/index.ts [hunk 12]
```diff
@@ -1,132 +1,231 @@
...
 // Copies binary data from buf to an svm_codec memory buffer
-export function copyToWasmBufferData(buf, data) {
-    let ptr = wasmBufferDataPtr(buf);
-    let memory = (codec!.exports as any).memory.buffer;
-    let view = new Uint8Array(memory);
+function copyToWasmBufferData(buf, data) {
+    const ptr = wasmBufferDataPtr(buf);
+    const memory = (codec.exports as any).memory.buffer;
+    const view = new Uint8Array(memory);
     view.set([...data], ptr);
```
Comment (brusherru) [inline src/index.ts:178]: How does it work?
It seems to me that it does nothing. What I missed?

File: src/index.ts [hunk 13]
```diff
@@ -1,132 +1,231 @@
...
+    return (codec.exports as any).wasm_buffer_data(buf);
 }
 
 // Copies binary data from buf to an svm_codec memory buffer
-export function copyToWasmBufferData(buf, data) {
-    let ptr = wasmBufferDataPtr(buf);
-    let memory = (codec!.exports as any).memory.buffer;
-    let view = new Uint8Array(memory);
+function copyToWasmBufferData(buf, data) {
+    const ptr = wasmBufferDataPtr(buf);
```
Comment (brusherru) [inline src/index.ts:175]: Shouldn't there be a check that the data fits in the buffer size?

File: src/index.ts [hunk 14]
```diff
@@ -1,132 +1,231 @@
...
+    const memory = (codec.exports as any).memory.buffer;
     const view = new Uint8Array(memory);
     return view.slice(ptr + offset, ptr + offset + length);
 }
 
 // Creates a wasm buffer that can be passed to an svm_codec instance methods from a provided json object
 // Returns the buffer
-export function newWasmBuffer(object: JSON) {
+function newWasmBuffer(object: any) {
     const objectStr = JSON.stringify(object);
```
Comment (brusherru) [inline src/index.ts:195]: Is it safe enough regarding further buffer allocation?

E.G.
1. Call `newWasmBufer({ some: 23 });`
2. It will be stringified to `'{"some":23}'` (length: 11)
3. Then the buffer length will be allocated for 11 bytes
4. Then we need to update such data to be `{ some: 100500 }`
5. It will not fit the buffer size

Probably I don't understand how exactly svm works (I didn't dive into it yet).
But it doesn't seem to me that buffer is immutable since there are functions that do something with the buffer. So if we want to store some related data in the same memory allocation — we had to allocate it smarter by taking into account the size of the type. But it also will require to store it in a different way (not just stringify, and then easily JSON.parse), which is much more complex 🤔

Comment (avive) [reply src/index.ts:195]: @YaronWittenstein ?

File: src/index.ts (hunk 13 above)
Comment (avive) [reply src/index.ts:175]: @YaronWittenstein

File: src/index.ts (hunk 12 above)
Comment (avive) [reply src/index.ts:178]: @YaronWittenstein ?

File: src/index.ts (hunk 10 above)
Comment (avive) [reply src/index.ts]: Good idea. I will make this change.

File: src/index.ts (hunk 9 above)
Comment (avive) [reply src/index.ts:142]: You are correct, I've been a bit lazy here. I will fix these nullable things.

File: src/index.ts (hunk 8 above)
Comment (avive) [reply src/index.ts:8]: The idea is that the library keeps the instance internally and wraps calls to it so there's no need to return that instance to clients. I think it is better to keep it simple like this and not add promises.

File: src/index.ts (hunk 11 above)
Comment (avive) [reply src/index.ts:229]: Good idea. Will make this change.

File: src/index.ts (hunk 8 above)
Comment (brusherru) [reply src/index.ts:8]: Keep simple — sure, so I don't really like an idea with a classic Singleton pattern.

However, if we'd provide a developer the only way to deal with the codec (by providing API from `init` function), we'll get rid of errors "not instantiated yet". And also you don't have to add checks for it in all internal functions.
Btw, if there should be only one wasm instance for the whole App in any case, it also can mutate some internal global `codec` variable (or check that it is already instantiated) and then provide the API.

In other cases, I can imagine that a lot of developers will wrap it in their own Singleton-like thing to deal with it without headache :)

E.G.
```
import initSvm from 'svm-codec-npm`;

(async () => {
  const wasmCode = await fetch(...);
  const api = await initSvm(wasmCode);
  api.encodeSpawn(...);
  ...
)();
```

File: src/index.ts (hunk 12 above)
Comment (YaronWittenstein) [reply src/index.ts:178]: This stuff is a bit low-level.

`wasmButterDataPtr` - it takes SVM Wasm buffer and returns a pointer to its `Data` Section.

https://github.com/spacemeshos/svm/blob/d91e8e76f81aa9319a4c4f109a6a9372ca0065c5/crates/codec/src/api/wasm/mod.rs#L89-L91

so we get that `ptr` points (Wasm Instance Memory offset) to the data held by `buf`.

Then we wrap the Wasm Instance Memory in a `Uint8Array`.
Lastly, we take that `Array` (i.e the Wasm Instance Memory) and copy the input param `data` of the function
to this array (starting at `offset = ptr`).

In other words, this function gets an input `data` and it overrides the Data held by the SVM Wasm Buffer given as input param `buf`

File: src/index.ts (hunk 13 above)
Comment (YaronWittenstein) [reply src/index.ts:175]: Seems that we don't really need it, since `copyToWasmBufferData` is an internal function.
But, it's still a good practice to add an assertion.

So @avive the `svm_codec.wasm` exposes `wasm_buffer_length`:
https://github.com/spacemeshos/svm/blob/d91e8e76f81aa9319a4c4f109a6a9372ca0065c5/crates/codec/src/api/wasm/mod.rs#L81

You can grab that and assert that the length of `data` fits into the `Data`-length pointed by SVM Wasm Buffer `buf`

File: src/index.ts (hunk 14 above)
Comment (YaronWittenstein) [reply src/index.ts:195]: All the buffers allocation are designed as one-time operations.

* You have an object (JSON)
* Then you allocate its byte-size as a buffer
* Then you copy the JSON into that buffer
* Lastly you call the Wasm instance
* Now the buffer should be freed 

All this mechanism is internal of course  (not part of the exported API)

File: tests/index.test.ts (hunk 5 above)
Comment (avive) [reply tests/index.test.ts:338]: opening an issue for this.

File: src/index.ts (hunk 8 above)
Comment (avive) [reply src/index.ts:8]: I think it is fine for this pr. Dev needs to init lib once by providing the wasm file and just calls functions on it. I'll open an issue about improving the interface based on your suggestion here.

File: src/index.ts (hunk 11 above)
Comment (avive) [reply src/index.ts:229]: that code was obsolete. Current code checks for expected error markers and throws an exception if a marker is unexpected.

File: src/index.ts (hunk 13 above)
Comment (avive) [reply src/index.ts:175]: I' like to have these kind improvements suggestions in follow up issues and prs. I will create them.
